src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+        let val1 = read_volatile(ptr1);
+        let val2 = read_volatile(ptr2);
+
+        if val1 != val2 {
+            info!("Test failed at {ptr1:?} compared to {ptr2:?}");
+            return Ok(MemtestOutcome::Fail(MemtestFailure::MismatchedValues {
+                address1: ptr1 as usize,
+                value1: val1,
+                address2: ptr2 as usize,
+                value2: val2,
+            }));
+        }
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+fn write_xor(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, val ^ read_volatile(ptr1));
+        write_volatile(ptr2, val ^ read_volatile(ptr2));
+    }
+}
+
+fn write_sub(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1).wrapping_sub(val));
+        write_volatile(ptr2, read_volatile(ptr2).wrapping_sub(val));
+    }
+}
+
+fn write_mul(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1).wrapping_mul(val));
+        write_volatile(ptr2, read_volatile(ptr2).wrapping_mul(val));
+    }
+}
+fn write_div(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    let val = if val == 0 { 1 } else { val };
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1) / val);
+        write_volatile(ptr2, read_volatile(ptr2) / val);
+    }
+}
+
+fn write_or(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1) | val);
+        write_volatile(ptr2, read_volatile(ptr2) | val);
+    }
+}
+
+fn write_and(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
+    unsafe {
+        write_volatile(ptr1, read_volatile(ptr1) & val);
+        write_volatile(ptr2, read_volatile(ptr2) & val);
+    }
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the result of adding a random value to the index of iteration
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_seq_inc(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestSeqInc");
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let expected_iter =
+        u64::try_from(half_count * 2).context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+
+    let value: usize = random();
+    for i in 0..half_count {
+        timeout_checker.check()?;
+        write_volatile(base_ptr.add(i), value.wrapping_add(i));
+        write_volatile(half_ptr.add(i), value.wrapping_add(i));
+    }
+    compare_regions(base_ptr, half_ptr, half_count, timeout_checker)
+}
+
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write to all bytes with either 0xff or 0x0 in an alternating pattern
```

Comment:
0xff and 0x0 make me think you're talking about 8 bits, but this is actually `usize` bits. Perhaps just say "all bit '0'" and "all bit '1'"?